<?xml version="1.0"?>
<launch>

<!--
Examples:
   F2M (default VO):
      $ roslaunch rtabmap_examples euroc_datasets.launch
      $ rosbag play -.-clock V1_01_easy.bag

      For MH sequences, we should set MH_seq to true because the ground truth source is different.
      $ roslaunch rtabmap_examples euroc_datasets.launch MH_seq:=true
      $ rosbag play -.-clock MH_01_easy.bag

   MSCKF (VIO):
      $ roslaunch rtabmap_examples euroc_datasets.launch args:="Odom/Strategy 8"
      $ rosbag play -.-clock V1_01_easy.bag

      We need to ignore the first 24 seconds for correct VIO initialization (drone should not move).
      $ roslaunch rtabmap_examples euroc_datasets.launch args:="Odom/Strategy 8" MH_seq:=true
      $ rosbag play -.-clock -s 24 MH_01_easy.bag

   OKVIS (VIO):
      $ roslaunch rtabmap_examples euroc_datasets.launch args:="Odom/Strategy 6 OdomOKVIS/ConfigPath ~/okvis/config/config_fpga_p2_euroc.yaml" MH_seq:=true raw_images_for_odom:=true
      $ rosbag play -.-clock MH_01_easy.bag

   VINS (VIO):
      $ roslaunch rtabmap_examples euroc_datasets.launch args:="Odom/Strategy 9 OdomVINS/ConfigPath ~/catkin_ws/src/VINS-Fusion/config/euroc/euroc_stereo_imu_config.yaml" MH_seq:=true raw_images_for_odom:=true
      $ rosbag play -.-clock MH_01_easy.bag

   VINS (VO):
      $ roslaunch rtabmap_examples euroc_datasets.launch args:="Odom/Strategy 9 OdomVINS/ConfigPath ~/catkin_ws/src/VINS-Fusion/config/euroc/euroc_stereo_config.yaml" MH_seq:=true raw_images_for_odom:=true
      $ rosbag play -.-clock MH_01_easy.bag
-->

<param name="use_sim_time" value="true"/>

<arg name="feature_type" default="6"/>
<arg name="gravity_opt" default="false"/> <!-- Rtabmap will use IMU data to add gravity constraints to graph -->

<arg name="args" default=""/>
<arg     if="$(arg gravity_opt)" name="common_args" default="-d --RGBD/CreateOccupancyGrid false --Odom/FeatureType $(arg feature_type) --Kp/DetectorStrategy $(arg feature_type) --Optimizer/GravitySigma 0.3 $(arg args)"/>
<arg unless="$(arg gravity_opt)" name="common_args" default="-d --RGBD/CreateOccupancyGrid false --Odom/FeatureType $(arg feature_type) --Kp/DetectorStrategy $(arg feature_type) $(arg args) "/>

<arg name="cfg" default=""/>
<arg name="raw_images_for_odom" default="false"/>
<arg name="record_ground_truth" default="false"/>
<arg name="rtabmap_viz" default="false"/>
<arg name="rviz" default="false"/>
<arg name="output_prefix" default="/tmp/rtabmap"/>

<node pkg="imu_complementary_filter" type="complementary_filter_node" name="imu_filter" output="screen">
   <remap from="imu/data_raw" to="/imu0"/>
   <param name="use_mag" value="false"/>
   <param name="world_frame" value="enu"/>
   <param name="publish_tf" value="false"/>
</node>

<!-- RTAB-Map -->
<include file="$(find rtabmap_odom)/launch/rtabmap.launch">
   <arg name="rtabmap_args" value="$(arg common_args) --Rtabmap/ImagesAlreadyRectified true"/>
   <arg name="odom_args" value="--Rtabmap/ImagesAlreadyRectified true"/>
   <arg name="left_image_topic" value="/multisense_sl/camera/left/image_raw"/>
   <arg name="left_camera_info_topic" value="/multisense_sl/camera/left/camera_info"/>
   <arg name="right_image_topic" value="/multisense_sl/camera/right/image_raw"/>
   <arg name="right_camera_info_topic" value="/multisense_sl/camera/right/camera_info"/>
   <arg name="stereo" value="true"/>
   <arg name="frame_id" value="base_footprint"/>
   <arg name="map_frame_id" value="rtabmap_map"/>
   <arg name="wait_for_transform" value="0.1"/>
   <arg if="$(arg record_ground_truth)" name="ground_truth_frame_id" value="world"/>
   <arg if="$(arg record_ground_truth)" name="ground_truth_base_frame_id" value="base_link_gt"/>
   <arg name="cfg" value="$(arg cfg)"/>
   <arg name="imu_topic" value="/imu/data"/>
   <arg name="rtabmap_viz" value="$(arg rtabmap_viz)"/>
   <arg name="rviz" value="$(arg rviz)"/>
   <arg name="wait_imu_to_init" value="true"/>
   <arg name="output_prefix" value="$(arg output_prefix)"/>
   <arg name="publish_tf_odom" value="false"/>
   <arg name="publish_tf_map" value="false"/>
   <arg name="database_path"  value=""/>
   <!-- <arg name="odom_topic" value="visual_odom"/> -->
</include>

</launch>
